### PR TITLE
Touch up and standardize README examples

### LINKED ISSUE #150: Add example detailing how to fire events
# There should be an example of how to fire events

Currently, there are no examples on how to fire regular or custom events in the README file. I believe it would be beneficial for people new to Haunted (or even web components) to have an example of how to fire events from within their component.

Related to this, on the README there are several examples showing different ways to create a component without speaking to the suttle differences that they create.

For example, there are a few components presented with arrow functions:

```js
const App = ({ name }) => {
 // ...
};

// and

const App = component(() => {
  // ...
});
```

But there's also a regular component function definition:

```js
function Counter() {
  // ...
}

customElements.define('my-counter', component(Counter));
```

**This should be standardized in all examples**. I consider the latter to be the most correct:

## Why does it matter that they're different?

While one might assume showing differently written examples to be harmless or even beneficial, this assumption fails to take into account and explain what either scenarios benefits or losses are.

One significant loss of arrow function components is that **you no longer have access to `this.dispatchEvent`** because of the implicit binding to `this` upon creation.

<br>

### This isn't only important for the component function definition

You might consider documenting that, when trying to do `this.dispatchEvent` or `this.anything` in a callback (that you might attach to a click event or the like), you have to bind the function manually or implicitly via an arrow function as simply creating a `function` definition within your component will bind the function to the element that called it which won't always be your web component.

As you can see in the below example, `fireIncrement` will be calling `dispatchEvent` on the `<button>` that was clicked on, not `Counter`.
```js
function Counter() {
  function fireIncrement() {
    console.log(this);
    // "<button>"
    this.dispatchEvent(/* ... */);
  }
  
  return html`<button @click=${fireIncrement}>Increment</button>`
}
```

However, this could also just be the developer's responsibility to know said information beforehand or it could be placed on a FAQ page of sorts.

### LINKED ISSUE #151: Standardization of examples in README
# The examples need to be standardized

The examples in the readme are all over the place in terms of being written the same way each time. There are numerous examples most written in a different way than the last. As explained in issue #150 this lack of standardization in examples does bring it's own issues related to firing events.

I propose that all component definitions be re-written to match a style similar to the first example:

```js
function Counter({ myProp }) {
  const [count, setCount] = useState(0);

  return html`
    <div id="count">${count}</div>
    <button type="button" @click=${() => setCount(count + 1)}>Increment</button>
  `;
}

customElements.define('my-counter', component(Counter));
```

### You don't have to get rid of examples of a different style of writing

I'm not saying that it shouldn't be allowed to show different ways to write components in the README but it should be explicitly written as an example of that style. Along with the example should be the style's benefits and drawbacks to help avoid developer confusion.

So you could take something like the following:

> The easiest way to create components is by importing component and creating a custom element like so:
> 
> ```js
> import { component } from 'haunted';
> import { html } from 'lit-html';
> 
> const App = ({ name }) => {
>   return html`Hello ${name}!`;
> };
> 
> customElements.define('my-app', component(App));
> 
> ```

And turn it into a more objective and informative example of how to write the component in a new manner:

> Another way to write your component is with the arrow function syntax (or fat arrow). It offers a more concise and clean way to write your code, however, it does come with a one drawback:
> 
> There's no way to use the `this` keyword to refer to the instance of your web component that is running.
> 
> ```js
> const App = ({ name }) => {
>   console.log(this); // => "undefined"
>   return html`Hello ${name}!`;
> };
> ```

### PR DESCRIPTION
**Pull request by ChrisBrownie55**
Fixes #150
Fixes #151

Changelog:

- Fixed grammar
- Touched up wording everywhere
- Made the list of imports for import example longer
- Reformatted components to use the `function` syntax as discussed in #151
  - Added an explanation as to why this syntax should be used
- Removed the irrelevant parts of examples (boilerplate, obvious imports)
- Elaborated in some examples to give a more detailed understanding of what's going on
- Added a section on using properties as they were only mentioned once before and in a confusing manner.
- Added a section on dispatching events as discussed in #150
- Added a small subsection explaining why virtual components don't have access to `this` and why fat arrow syntax is encouraged when writing virtual components
- Added a section explaining dependencies and how they work with `useEffect`
- Added a section for `useCallback`
- Now using an empty array for `useEffect` dependencies in "Cleaning up side-effects" example to avoid rerunning the side-effect and it's clean up on every rerender
- Now using `parseInt` instead of `Number` in example
- Explained how and when to run a side-effect on "mount"
- Code formatting

If there's anything in here that looks like it needs to be changed/fixed let me know 👍 
Thank you for your time.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1,31 +1,32 @@
-# Haunted 🦇 🎃
+🦇 🎃
```
Comment (matthewp) [inline readme.md]: Why ?

File: readme.md [hunk 2]
```diff
@@ -1,31 +1,32 @@
...
 
 [![npm](https://img.shields.io/npm/dt/haunted)](https://npm.im/haunted)
 [![npm](https://img.shields.io/npm/v/haunted)](https://npm.im/haunted)
 
 React's Hooks API but for standard web components and [lit-html](https://lit-html.polymer-project.org/) or [hyperHTML](https://codepen.io/WebReflection/pen/pxXrdy?editors=0010).
 
 ```html
-<!doctype html>
```
Comment (matthewp) [inline readme.md]: This is usually wanted, why remove this?

**Comment by matthewp** [comment]
Thanks for taking this on! I have a couple of comments. Also can you squash your commits into 1?

File: readme.md (hunk 1 above)
Comment (ChrisBrownie55) [reply readme.md]: I didn't even realize that happened, not sure how it did, I'll fix that right away 👍

File: readme.md (hunk 2 above)
Comment (ChrisBrownie55) [reply readme.md]: It's usually wanted for a full web page, as are things like `<head>`, `<title>`, and `<body>` but they're unnecessary for just the bare example that's being given here

Comment (ChrisBrownie55) [reply readme.md]: I can add it back in if the example is better with it.

**Comment by ChrisBrownie55** [comment]
Thank you! I'll get started on fixing the aforementioned issues and then squashing the commits 👍

**Comment by ChrisBrownie55** [comment]
You should be able to squash the commits automatically when you merge the pull request:

https://help.github.com/en/github/administering-a-repository/configuring-commit-squashing-for-pull-requests

I believe it looks like the following when you can squash and merge:

image

**Comment by matthewp** [comment]
Ah yeah, I forgot about this feature. Thanks!

**Pull request merged**